src/main/scala/lantern/TensorDifferentiation.scala

```diff
@@ -2993,7 +3001,31 @@ trait TensorDslCublas extends TensorDsl with GPUOps {
     override def nllLoss_grad(input: TensorR, res: TensorR, target: Rep[Array[Int]]): Unit = {
       input.moveToCPU()
       res.moveToCPU()
+
+      // TODO: Use `withBackend` when implemented.
+      val tmp = backend
+      backend = BackendCPU()
```

Comment:
Temporarily changing `backend` to `BackendCPU()` is important.

Otherwise, calling `BackendCPU().xxx` methods is bug-prone in general because `xxx` may call methods defined on `Tensor`. For instance, if `xxx` calls `tensor + tensor`, the `+` will dispatch to the current `backend` value, which may not be `BackendCPU`.

One solution is to avoid using any `Tensor`-defined methods in `Backend` functions. This means that we should not write `t + t`, but rather `+(t, t)` (where `+` is the method defined on `Backend`).

Reply:
I agree. When Backend defines a function, it should be specific to a given backend (CPU or GPU), and it should be op(t, t) shape. Also for methods in Tensor class or TensorR class, they should always call backend.op(this, that, ...), so that methods in Tensor(R) class can always despatch according to the backend. 

We should keep this rule in mind, and fix stuff accordingly, I think.